core/utxo/txhash/encode.go

```
@@ -0,0 +1,183 @@
+package txhash
+
+import (
+	"crypto/sha256"
+	"encoding/binary"
+	"io"
+	"log"
+	"sort"
+
+	"github.com/xuperchain/xuperchain/core/pb"
+)
+
+type encoder struct {
+	intbuf [8]byte
+	w      io.Writer
+}
+
+func newEncoder(w io.Writer) *encoder {
+	return &encoder{
+		w: w,
+	}
+}
+
+func (e *encoder) EncodeInt64(x int64) {
+	buf := e.intbuf[:8]
+	binary.BigEndian.PutUint64(buf, uint64(x))
+	e.w.Write(buf)
+}
+
+func (e *encoder) EncodeString(s string) {
+	e.EncodeInt64(int64(len(s)))
+	if len(s) == 0 {
+		return
+	}
+	io.WriteString(e.w, s)
+}
+
+func (e *encoder) EncodeBytes(s []byte) {
+	e.EncodeInt64(int64(len(s)))
+	if len(s) == 0 {
+		return
+	}
+	e.w.Write(s)
+}
+
+func (e *encoder) EncodeMap(m map[string][]byte) {
+	length := len(m)
+	e.EncodeInt64(int64(length))
+	if length == 0 {
+		return
+	}
+	keys := make([]string, 0, len(m))
+	for key := range m {
+		keys = append(keys, key)
+	}
+	sort.Strings(keys)
+
+	e.Encode(len(m))
+	for _, key := range keys {
+		e.EncodeString(key)
+		e.EncodeBytes(m[key])
+	}
+}
+
+func (e *encoder) Encode(x interface{}) {
+	switch v := x.(type) {
+	case bool:
+		if v {
+			e.EncodeInt64(1)
+		} else {
+			e.EncodeInt64(0)
+		}
+	case int:
+		e.EncodeInt64(int64(v))
+	case int32:
+		e.EncodeInt64(int64(v))
+	case int64:
+		e.EncodeInt64(v)
+	case string:
+		e.EncodeString(v)
+	case []byte:
+		e.EncodeBytes(v)
+	case map[string][]byte:
+		e.EncodeMap(v)
+	default:
+		log.Panicf("not supported type:%T", x)
```

Why use Panicf here? It's better to throw an exception?